File: amazon/675_cut_off_trees_for_golf_event.py


class Solution:
    def bfs(self, forest: List[List[int]], sr: int, sc: int, tr: int, tc: int) -> int :
        if sr == tr and sc == tc:
            return 0
        m,n = len(forest),len(forest[0])
        queue = deque([(sr, sc, 0)]) # (row,col,steps)
        visited = set((sr, sc))
        directions=[(-1, 0),(1,0),(0,-1),(0,1)] #up, down, left, right

        while queue:
            r,c, steps = queue.popleft()
            for dr, dc in directions:
                nr, nc = r + dr, c + dc
                if 0 <= nr < m and 0 <= nc < n and (nr,nc) not in visited and forest[nr][nc] > 0:
                    if nr == tr and nc == tc:
                        return steps + 1
                    queue.append((nr,nc,steps+1))
                    visited.add((nr, nc))

        return -1

    def cutOffTree(self, forest: List[List[int]]) -> int:
        m,n = len(forest),len(forest[0])
        #collect all trees with their heights and positions
        trees = sorted((h,r,c) for r, row in enumerate(forest) for c, h in enumerate(row) if h > 1)
        # start from(0, 0)
        sr, sc = 0, 0
        total_steps = 0

        for _, tr, tc in trees:
            steps = self.bfs(forest, sr, sc, tr, tc)
            if steps == -1:
                return -1
            total_steps += steps
            #move the starting point to the current tree;s position
            sr, sc = tr, tc

        return total_steps
        

File: amazon/675_cut_off_trees_for_golf_event.java
import java.util.ArrayList;
import java.util.Queue;

class Solution {

    private static final int[][] DIRECTIONS = {{-1,0},{1,0},{0,-1},{0,1}};
    private int bfs(List<List<Integer>> forest,int sr, int sc, int tr, int tc){
        int m = forest.size();
        int n = forest.get(0).size();
        boolean[][] visited = new boolean[m][n];
        Queue<int[]> queue = new LinkedList<>();
        queue.add(new int[]{sr,sc,0});
        visited[sr][sc] = true;

        while(!queue.isEmpty()){
            int[] cur = queue.poll();
            int r = cur[0], c = cur[1], steps = cur[2];

            if(r == tr && c == tc) return steps;

            for(int[] dir : DIRECTIONS){
                int nr = r + dir[0], nc = c + dir[1];
                if(nr >= 0 && nr < m && nc >= 0 && nc < n && !visited[nr][nc] && forest.get(nr).get(nc) > 0){
                    queue.add(new int[]{nr, nc, steps +1});
                    visited[nr][nc] = true;
                }
            }
        }
        return -1;
    }

    public int cutOffTree(List<List<Integer>> forest) {
        int m = forest.size();
        int n = forest.get(0).size();

        List<int[]> trees = new ArrayList<>();
        for(int i = 0; i < m ; i++){
            for(int j =0; j < n; j++){
                int height = forest.get(i).get(j);
                if(height > 1){
                    trees.add(new int[]{height, i, j});
                }
            }
        }
        Collections.sort(trees, Comparator.comparingInt(a->a[0]));
        int totalSteps = 0;
        int sr = 0, sc = 0;

        for(int[] tree: trees){
            int tr = tree[1], tc = tree[2];
            int steps = bfs(forest, sr, sc, tr, tc); 
            if(steps == -1) return -1;
            totalSteps += steps;
            sr = tr;
            sc = tc;
        }
        return totalSteps;
    }
}